Tests/kaas/k8s-version-recency/k8s-version-recency-check.py

```diff
@@ -0,0 +1,339 @@
+        else:
+            return False
+
+    def get_version(self):
+        return f"{self.major}.{self.minor}.{self.patch}"
+
+
+def diff_months(date1, date2):
+    return abs((date1.year - date2.year) * 12 + date1.month - date2.month)
+
+
+def diff_weeks(date1, date2):
+    day1 = (date1 - datetime.timedelta(days=date1.weekday()))
+    day2 = (date2 - datetime.timedelta(days=date2.weekday()))
+    diff = day2 - day1
+    return abs((diff.days / 7) + math.ceil(diff.seconds / 86400))
+
+
+def diff_days(date1, date2):
+    delta = date1 - date2
+    return abs(delta.days)
+
+
+def check_cve_versions():
+
+    # CVE fix versions
+    cfvs = list()
+
+    # Request latest version
+    cve_list = requests.get("https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json",
+                            headers={"Accept": "application/json"}).json()
+
+    for cve in cve_list['items']:
+        cveid = cve['external_url'].split("=")[-1]
+        cve_data = requests.get(f"https://cveawg.mitre.org/api/cve/{cveid}",
+                                headers={"Accept": "application/json"}).json()
+        try:
+            cve_cna = cve_data['containers']['cna']
+            cve_metrics = cve_cna['metrics'][0]
+            key = None
+            for k in cve_metrics.keys():
+                if re.search(r'[cC][vV][sS]{1,2}V\d', k):
+                    key = k
+                    break
+
+            if cve_metrics[key]['baseScore'] >= CVE_SEVERITY:
+                for v in cve_cna['affected'][0]['versions']:
+                    vi = K8sVersionInfo()
+                    if 'lessThanOrEqual' in v.keys():
+                        vi.extract_version(v['lessThanOrEqual'].strip("v"))
+                        vi.patch += 1
+                    elif 'lessThan' in v.keys():
+                        vi.extract_version(v['lessThan'].strip("v"))
+                    if vi not in cfvs:
+                        cfvs.append(vi)
+        except KeyError as e:
+            logger.debug(f"They key {e} couldn't be found in the CVE json data for CVE {cveid}.")
+            pass
+
+    return cfvs
+
+
+async def get_k8s_cluster_version(kubeconfig):
+
+    await kubernetes_asyncio.config.load_kube_config(kubeconfig)
+
+    async with kubernetes_asyncio.client.ApiClient() as api:
+        version_api = kubernetes_asyncio.client.VersionApi(api)
+        ret = await version_api.get_code()
+
+        version = K8sVersionInfo()
+        version.extract_version(ret.git_version.strip("v"))
+        version.date = datetime.datetime.strptime(ret.build_date, '%Y-%m-%dT%H:%M:%SZ')
+
+        with open(kubeconfig, "r") as fd:
+            cluster_config = yaml.safe_load(fd)
+            fd.close()
+
+        return version, cluster_config['contexts'][0]['name']
+
+
+def get_k8s_versions(token, cve_versions=None):
+
+    k8s_versions = []
+
+    github_headers = {
+        "Accept": "application/vnd.github+json",
+        "Authorization": f"Bearer {token}",
```

Comment:
I'm wondering if this token business is really necessary. Which of the subsequent requests really need it? Can you add some more commentary about that? I'm asking because the token complicates matters, the information should be publicly available, and at least the "latest" page does indeed work straight from my browser without any special headers.

Reply:
But it seems to work without the token...
```
$ curl https://api.github.com/repos/kubernetes/kubernetes/releases/latest
{
  "url": "https://api.github.com/repos/kubernetes/kubernetes/releases/108644552",
  "assets_url": "https://api.github.com/repos/kubernetes/kubernetes/releases/108644552/assets",
  "upload_url": "https://uploads.github.com/repos/kubernetes/kubernetes/releases/108644552/assets{?name,label}",
  "html_url": "https://github.com/kubernetes/kubernetes/releases/tag/v1.27.3",
  "id": 108644552,
  "author": {
    "login": "k8s-release-robot",
    "id": 33505452,
    "node_id": "MDQ6VXNlcjMzNTA1NDUy",
    "avatar_url": "https://avatars.githubusercontent.com/u/33505452?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/k8s-release-robot",
    "html_url": "https://github.com/k8s-release-robot",
    "followers_url": "https://api.github.com/users/k8s-release-robot/followers",
    "following_url": "https://api.github.com/users/k8s-release-robot/following{/other_user}",
    "gists_url": "https://api.github.com/users/k8s-release-robot/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/k8s-release-robot/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/k8s-release-robot/subscriptions",
    "organizations_url": "https://api.github.com/users/k8s-release-robot/orgs",
    "repos_url": "https://api.github.com/users/k8s-release-robot/repos",
    "events_url": "https://api.github.com/users/k8s-release-robot/events{/privacy}",
    "received_events_url": "https://api.github.com/users/k8s-release-robot/received_events",
    "type": "User",
    "site_admin": false
  },
  "node_id": "RE_kwDOAToIks4GecjI",
  "tag_name": "v1.27.3",
  "target_commitish": "master",
  "name": "Kubernetes v1.27.3",
  "draft": false,
  "prerelease": false,
  "created_at": "2023-06-14T09:47:38Z",
  "published_at": "2023-06-14T21:36:39Z",
  "assets": [

  ],
  "tarball_url": "https://api.github.com/repos/kubernetes/kubernetes/tarball/v1.27.3",
  "zipball_url": "https://api.github.com/repos/kubernetes/kubernetes/zipball/v1.27.3",
  "body": "\nSee [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce). Additional binary downloads are linked in the [CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md).\n\nSee [the CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md) for more details.\n\n\n\n",
  "reactions": {
    "url": "https://api.github.com/repos/kubernetes/kubernetes/releases/108644552/reactions",
    "total_count": 45,
    "+1": 13,
    "-1": 0,
    "laugh": 2,
    "hooray": 15,
    "confused": 0,
    "heart": 4,
    "rocket": 5,
    "eyes": 6
  }
}
```

Reply:
Interesting, I got an error (I think 403) last time I tried that and then the [Docs](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases) brought me to the conclusion that I needed a token.
But the text also says only the 'draft' releases need a token, so maybe I did something wrong last time.

Then I will change that up.

Reply:
If that works, great. Then please remove any mention of the token from the usage and getopt stuff.